File: src/main/java/dev/slimevr/autobone/errors/HeightError.java
package dev.slimevr.autobone.errors;

import com.jme3.math.FastMath;

import dev.slimevr.autobone.AutoBoneTrainingStep;


public class HeightError implements IAutoBoneError {
	@Override
	public float getStepError(AutoBoneTrainingStep trainingStep) throws AutoBoneException {
		return getHeightError(
			trainingStep.getCurrentHeight(),
			trainingStep.getTargetHeight()
		);
	}

	public float getHeightError(float currentHeight, float targetHeight) {
		return FastMath.abs(targetHeight - currentHeight);
	}
}

File: server/core/src/main/java/dev/slimevr/autobone/errors/HeightError.kt
package dev.slimevr.autobone.errors

import dev.slimevr.autobone.AutoBoneStep
import dev.slimevr.autobone.PoseFrameStep
import kotlin.math.*

// The difference from the current height to the target height
class HeightError : IAutoBoneError {
	@Throws(AutoBoneException::class)
	override fun getStepError(step: PoseFrameStep<AutoBoneStep>): Float = getHeightError(
		step.data.hmdHeight,
		step.data.targetHmdHeight,
	)

	fun getHeightError(currentHeight: Float, targetHeight: Float): Float = abs(targetHeight - currentHeight)
}
